pkg/datapath/linux/ipsec/ipsec_linux.go

```diff
@@ -740,13 +746,64 @@ func ipsecDeleteXfrmState(nodeID uint16) {
 		scopedLog.WithError(err).Warning("Failed to list XFRM states for deletion")
 		return
 	}
+	xfrmStatesToDelete := []netlink.XfrmState{}
+	oldXfrmInStates := map[oldXfrmStateKey]netlink.XfrmState{}
 	for _, s := range xfrmStateList {
 		if matchesOnNodeID(s.Mark) && ipsec.GetNodeIDFromXfrmMark(s.Mark) == nodeID {
-			if err := netlink.XfrmStateDel(&s); err != nil {
-				scopedLog.WithError(err).Warning("Failed to delete XFRM state")
+			xfrmStatesToDelete = append(xfrmStatesToDelete, s)
+		}
+		if xfrmMarkEqual(s.Mark, oldXFRMInMark) {
+			key := oldXfrmStateKey{
+				Spi: s.Spi,
+				Dst: [16]byte(s.Dst.To16()),
 			}
+			oldXfrmInStates[key] = s
 		}
 	}
+	for _, s := range xfrmStatesToDelete {
+		key := oldXfrmStateKey{
+			Spi: s.Spi,
+			Dst: [16]byte(s.Dst.To16()),
+		}
+		var oldXfrmInState *netlink.XfrmState = nil
+		old, ok := oldXfrmInStates[key]
+		if ok {
+			oldXfrmInState = &old
+		}
+		if err := safeDeleteXfrmState(&s, oldXfrmInState); err != nil {
+			scopedLog.WithError(err).Warning("Failed to delete XFRM state")
+		}
+	}
+}
+
+// safeDeleteXfrmState deletes the given XFRM state. Specifically, if the
+// state is to catch ingress traffic marked with nodeID (0xXXXX0d00), we
+// temporarily remove the old XFRM state that matches 0xd00/0xf00. This is to
+// workaround a kernel issue that prevents us from deleting a specific XFRM
+// state (e.g. catching 0xXXXX0d00/0xffff0f00) when there is also a general
+// xfrm state (e.g. catching 0xd00/0xf00). When both XFRM states coexist,
+// kernel deletes the general XFRM state instead of the specific one, even if
+// the deleting request is for the specific one.
+func safeDeleteXfrmState(state *netlink.XfrmState, oldState *netlink.XfrmState) (err error) {
+	if getDirFromXfrmMark(state.Mark) == dirIngress && ipsec.GetNodeIDFromXfrmMark(state.Mark) != 0 && oldState != nil {
+
+		scopedLog := log.WithFields(logrus.Fields{
+			logfields.SPI:              state.Spi,
+			logfields.SourceIP:         state.Src,
+			logfields.DestinationIP:    state.Dst,
+			logfields.TrafficDirection: getDirFromXfrmMark(state.Mark),
+			logfields.NodeID:           getNodeIDAsHexFromXfrmMark(state.Mark),
+		})
+
+		err, deferFn := xfrmTemporarilyRemoveState(scopedLog, *oldState, string(dirIngress))
+		if err != nil {
+			scopedLog.WithError(err).Errorf("Failed to remove old XFRM %s state %s", string(dirIngress), oldState.String())
+		} else {
+			defer deferFn()
+		}
+	}
```

Comment:
In the upstream patch, we would `return err` here on error. Is that diff intentional?

Reply:
You are right, I should return err if err != nil. I didn't return because I referred to the diff between xfrmStateReplace in 1.14 https://github.com/cilium/cilium/blob/v1.14/pkg/datapath/linux/ipsec/ipsec_linux.go#L369 and xfrmStateReplace in main https://github.com/cilium/cilium/blob/main/pkg/datapath/linux/ipsec/ipsec_linux.go#L381.